dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,313 @@
+  if (has_channels) {
+    return dim == ndim - 1;
+  } else {
+    return dim == ndim;
+  }
+}
+template <int dim, int ndim, bool has_channels>
+using is_convolution_inner = std::enable_if_t<is_convolution_inner_loop(dim, ndim, has_channels)>;
+
+template <int dim, int ndim, bool has_channels>
+using is_convolution_outer = std::enable_if_t<!is_convolution_inner_loop(dim, ndim, has_channels)>;
+
+// we're in channel dim
+template <int axis, bool has_channels, int dim = 0, typename Out, typename In, typename W, int ndim>
+is_convolution_inner<dim, ndim, has_channels> ConvolutionCpuImpl(
+    Out* out, const In* in, const W* window, const TensorShape<ndim>& shape,
+    const TensorShape<ndim>& strides, int d, int64_t offset, span<const In> border_fill,
+    In* input_window_buffer, span<W> pixel_tmp, W scale = 1) {
+  auto pixel_stride = strides[axis];
+  auto axis_size = shape[axis];
+  auto num_channels = has_channels ? shape[ndim - 1] : 1;  // channel-last is assumed
+  int r = (d - 1) / 2;                                     // radius = (diameter - 1) / 2
+  // offset <- start of current axis
+  auto* out_ptr = out + offset;
+  auto* in_ptr = in + offset;
+
+  CyclicPixelWrapper<In, has_channels> input_window(input_window_buffer, d, num_channels);
+
+  int in_idx = -r, out_idx = 0;
+  for (in_idx = -r; in_idx < 0; in_idx++) {
+    load_pixel_with_border(input_window, in_ptr, in_idx, pixel_stride, axis_size, border_fill);
+  }
+  if (r < axis_size) {
+    // we load the window without the last element
+    for (; in_idx < r; in_idx++) {
+      load_pixel_no_border(input_window, in_ptr, in_idx, pixel_stride);
+    }
+    for (; out_idx < axis_size - r; out_idx++, in_idx++) {
+      // we load last element of the input window corresponding to the out_idx
+      load_pixel_no_border(input_window, in_ptr, in_idx, pixel_stride);
+      // we have both windows as almost-contiguous buffers
+      input_window.CalculateDot(pixel_tmp.data(), window);
+      for (int c = 0; c < num_channels; c++) {
+        out_ptr[out_idx * pixel_stride + c] = ConvertSat<Out>(pixel_tmp[c] * scale);
+      }
+      // remove one pixel, to make space for next out_idx and in_idx
+      input_window.PopPixel();
+    }
+  } else {
+    // we need to load the rest of the window, just handle all with border condition for simplicity
+    for (; in_idx < r; in_idx++) {
+      load_pixel_with_border(input_window, in_ptr, in_idx, pixel_stride, axis_size, border_fill);
+    }
+  }
+  // we need write out the rest of the outputs, the input window is full of data
+  for (; out_idx < axis_size; out_idx++, in_idx++) {
+    load_pixel_with_border(input_window, in_ptr, in_idx, pixel_stride, axis_size, border_fill);
+    input_window.CalculateDot(pixel_tmp.data(), window);
+    for (int c = 0; c < num_channels; c++) {
+      out_ptr[out_idx * pixel_stride + c] = ConvertSat<Out>(pixel_tmp[c] * scale);
+    }
+    input_window.PopPixel();
```

Comment:
Maybe you should not call this Pop directly and let the load_pixel* handle it?

Reply:
I know when to call Pop but the load_pixel doesn't, so I would need to place additional logic there.